### ISSUE TITLE
Add Vulkan renderer

### ISSUE THREAD

**Issue by Marqin** [urho3d/Urho3D#1201]
Hi!

It would be nice to support Vulkan API, since it's [already released](https://www.khronos.org/vulkan/).

**Comment by fire** [comment]
Is anyone familiar with the work needed? I have some interest. Feel free to chat with me in the #urho3d irc channel.

**Comment by cadaver** [comment]
See also the Vulkan thread on the forum. http://urho3d.prophpbb.com/topic1861.html

**Comment by boberfly** [comment]
Too bad all the drivers for the hardware are not working very well, at least for me on Linux. I have a laptop with an Ivy Bridge Intel GPU and an Nvidia Geforce GTX 765m, both of which have problems running the demos. I can render a triangle on the Geforce...

Having said that, I'm still structuring a backend here and there when I get time and will test it once the drivers are working. To really take advantage of the API, Urho's renderer will need some extensive refactoring.

**Comment by OvermindDL1** [comment]
On Fri, Mar 4, 2016 at 12:14 PM, Alex Fuller <notifications@github.com>
wrote:

> Too bad all the drivers for the hardware are not working very well, at
> least for me on Linux. I have a laptop with an Ivy Bridge Intel GPU and an
> Nvidia Geforce GTX 765m, both of which have problems running the demos. I
> can render a triangle on the Geforce...
>
> Having said that, I'm still structuring a backend here and there when I
> get time and will test it once the drivers are working. To really take
> advantage of the API, Urho's renderer will need some extensive refactoring.
>
Might be a good time to refactor Urho3D in any case then with perhaps a
newer name.  Urho3D follows a very Unity'ish EC programming style instead
of proper ECS, which loses a lot of the usefulness of ECS.  Proper ECS will
also make it *significantly* easier to support newer platforms in the
future, like Vulkan would have been if proper ECS was followed.  Most of
the code in Urho3D as-is would be reusable with such a major refactoring,
but the overall programming style would change.

An example of an ECS library (just a library, can be used with any
renderer, any project, anywhere) in C++ is entityx[0], and although it is
quite efficient, there are better ways of doing it as ECS is a subtype of
Dataflow and is supposed to have the advantages of Dataflow with more.

Right now Urho3D is just EC, like Unity.  It combines the Components and
Systems into one thing that causes a performance drop in acquiring the
components, and by default (though I have a custom factory for mine to fix
this) Urho3D allocates them on the heap instead of in a cache-friendly
manner.  For note, Systems in Urho3D are like Systems in ECS that do not
iterate over the components, and right now in Urho3D if you want to iterate
over the components you either have to do it very slowly by iterating over
near everything or have to make custom View adaptors to listen to
(de)registration events, which thanks to the components being scattered
across memory is inefficient regardless.

A nice side effect of such a refactor would be that Components become pure
data storage, and there are two main ways to do this efficiently, either
have main component storage (so instead of
someEntity.getComponent<HealthComponent>() you would do
world.getComponentStore<HealthComponent>().getHealth(someEntity), although
that can have wrappers for ease of use like the above) and the
HealthComponent is now something like:
```
struct HealthComponent {
  int hp;
}
```
But rather something like:
```
class HealthComponent {
  typedef int Health;
  Bag<Health> healths_; // Bag is whatever is appropriate for storage,
maybe an std::vector for *very* common components like Position, maybe an
ordered set for common things, maybe just an unordered set of some sort for
rare things or whatever is appropriate for the use case
  public:  //getters/setters...
}
```
Or you can make them 'read-only' during normal processing and store any
changes for storage at the end of the tick, and although this has an
additional (small in near every case) memory overhead it does allow for
arbitrary parallelization of the Systems.  You can still get
parallelization of the Systems without such a lockstep (dataflow) pattern
but you have to build a tree of the systems in order to properly order them
so Systems that access the same components in a possible-changing-way do
not run concurrently, which is still possible, but also breaks up a lot of
possible parallelization opportunities.

Systems in this style are something like:
```
class ApplyDamageSystem
  :public WhateverBaseSystemType<
    ComponentView.all<DamageComponent, HealthComponent>
  > { // this could be passed in the constructor as an object instead of as
a template, but you can still build Views manually and this allows for more
optimizations, like building a parallelization tree easier
 // The base class would have a view_ member or something that holds the
view, and the view gets updated with a list of the entities that fulfill
the view requirements, in this case having both a DamageComponent and a
HealthComponent, but you can build views other ways too, then of course you
can call an update method on the system or whatever custom stuff you want
 //  The view_ storage is usually just an ordered_set of entity types,
though you can make custom ones to store direct pointers to components or
whatever for more efficiency, and it is an ordered set as lookups of entity
components in order should be cache friendly as most should be stored in
order if done right
}
```
You could, for example, having during a collision (handled by another
system) apply a DamageComponent to the necessary entities, which would then
be picked up by this system on the next run (either next tick if ordered
before, this tick if ordered after, or next tick if mutations are applied
at end-of-tick, depending on the style the engine is made), which could
then run over it and apply the necessary transformation on the
HealthComponent (which if it is a const/mutation based system then during
the mutating step it could apply a KilledComponent or so if health<=0 via
the HealthComponent mutator itself instead of needing to worry about the
Systems that use the HealthComponent from doing it themselves).

Sorry for the email length, I am stopping, I could keep going on about how
to do this design properly for a decent book, there is so much incorrect
information online about it everywhere...

[0]: https://github.com/alecthomas/entityx

**Comment by boberfly** [comment]
Hey OvermindDL1,

Yeah I've done a bit of research on the matter before, and certainly this would be a more cache friendly approach, and would also perhaps improve things on the renderer side. It sounds like a huge undertaking though and would break backwards-compatibility or anyone relying on Urho3D for their projects. Perhaps Urho3D 2.0? You need the man-power and time for it...

The refactor of the rendering side which I was implying isn't necessarily about the component system though. It was more or less the higher-level parts of the renderer that sit on top of the API abstraction. One possible redesign could be to make it work in the most optimal way that Vulkan or DX12 wants, and then make emulator parts for the other APIs (which may impose a slight performance penalty on them unfortunately, but that's speculation until you do it in practice).

**Comment by OvermindDL1** [comment]
On Fri, Mar 4, 2016 at 1:37 PM, Alex Fuller <notifications@github.com>
wrote:

> Hey OvermindDL1,
>
> Yeah I've done a bit of research on the matter before, and certainly this
> would be a more cache friendly approach, and would also perhaps improve
> things on the renderer side. It sounds like a huge undertaking though and
> would break backwards-compatibility or anyone relying on Urho3D for their
> projects. Perhaps Urho3D 2.0? You need the man-power and time for it...
>
True, and if anyone starts such a project then point me to it and I'll
help.

**Comment by boberfly** [comment]
Kinda going off-topic, but I'd like to rename the humble 'Node' to something like 'Entity' though, so the Node name can be used for stuff like blueprint-inspired nodes :) and the TransformComponent would be stuff that needs/does transforms, physics/graphics/etc.

Cadaver uses Turso3D to do experiments, I'm not sure though if it's working like ECS:
https://github.com/cadaver/turso3d

**Comment by cadaver** [comment]
Turso3D uses a componentless approach; there is just a hierarchy of different kinds of scene nodes, which can be composed, for example a Mesh as a child of a RigidBody, or a Script node affecting its parent node in some manner (not all nodes need to be spatial)

I did some experiments on memory access efficiency but so far its quite similar to Urho3D.

I was mostly interested in increasing the efficiency of rendering CPU side, where I believe it's hard to come up with a good memory use mechanism, because to process the scene fully you end up accessing a lot of data no matter how you order it (e.g. world matrices, bounds, materials/passes, render queues, render queue manipulation to auto-find instanceable objects).

Ogre2+ spent a lot of energy into optimizing the memory access of scene graph processing & rendering but the last I measured Urho3D still outperformed it in simple "torture" tests like the HugeObjectCount sample, while Turso3D being even a bit faster due to optimizations and simpler operation.

A pure ECS approach would no doubt speed up some applications, whose processing can be clearly divided to systems. I would think it would be superior if you code everything yourself, including the physics engine. Otherwise you'll be still referring a bit clumsily to outside object instances, such as Bullet's rigidbodies. Personally I'm not a fan of ECS because when applied rigidly, as it always requires creating a system to do the processing. If you were to do a pure C++ engine I could see its appeal, however combining C++ systems and scripted systems, and C++ components and scripted components would get difficult and likely lose the performance gain.

In other words, be my guest in making an ECS project if you have the inspiration, Urho3D is unlikely to go into that direction at least for the time being. Note that even in Urho3D nothing prevents you processing components manually in your own systems, which can reduce lookup costs (eg. a NPCManager system which knows of all NPC components), though the memory allocation for the components (if you use the standard approach of Node's CreateComponent) is naturally not optimal.

**Comment by boberfly** [comment]
Thanks for the insight Cadaver. I'm still leaning towards special-case stuff like crowds and particle systems for stuff that needs brute-force, and they can have their own managers that hold the components like you said (with some cache-friendly data layout), and the hot-path would have functions in that manager which takes in that with an offset and mask (to determine which components need calculation), preferably in WorkItems for threading in tiles.

Now back to Vulkan :) I'm guessing cadaver after the octree process to determine what to draw, you create batches for the renderer to ingest. This is threaded from what I see, I'm wondering if the steps straight after could be threaded also to write directly to the command buffers of the API. There's still that notion that there's a global renderer state unfortunately which complicates things a bit. I did read somewhere that a possible abstraction approach is to have a struct which has all of the state needed to invoke a draw call, so in a sense it could appear 'stateless' to the older APIs while making the newer API implementations straightforward, by just holding pipeline/descriptors/etc. and chances to cache command buffers which don't change by a dirty flag of sorts. Obviously some kind of system to sort these structs in a way that you reduce state changes would need to be made to not cause performance regressions, I'm guessing the batch class already does this, but I need to get my head around it a bit more...

And stuff like blendshapes or things which modify vertex buffers would have their own dynamic memory pools on each thread... Lots to consider, talking out aloud here.

**Comment by boberfly** [comment]
Putting this here so I don't forget about it:
https://github.com/powervr-graphics/Native_SDK/tree/4.1/Framework/PVRApi

A nice reference for abstracting Vulkan/GLES/others.

**Comment by fire** [comment]
Here's a from scratch vulkan render that is based on glfw and outputs imgui. https://github.com/ocornut/imgui/blob/master/examples/vulkan_example/main.cpp

Not sure how helpful it is, but the example is small.

**Comment by boberfly** [comment]
@fire cheers, looks like a nice hello world of Vulkan. Similar to the init I've sort of made in Urho based off the samples I've found here and there.

**Comment by boberfly** [comment]
@cadaver this SDK from PowerVR is actually quite interesting and liberally licensed. They seem to use [this](https://github.com/cameron314/concurrentqueue) for threading too which I didn't know about until now, but the benchmarks against TBB are promising. The code is not directly portable to Urho3D, but there's lots of inspiration for the abstraction & how they did the gnome horde demo rendering. 

If anyone wants an abstraction layer over the built-in cameras for IOS and Android, PVRCamera gives you that (which I may use for yet another side-project that I may or may not finish). I could imagine that being used for AR stuff on mobile.

**Comment by Kavignon** [comment]
This would be a nice feature inside urho. What are the actual update of the issue ?

**Comment by Kavignon** [comment]
@cadaver what's the current status on the issue? If I were to start working on this, what should I read to be ready to integrate this smoothly? What spikes should be done ? I'd really love to be helpful on the matter at hands and fully integrate this feature in Urho

**Comment by cadaver** [comment]
I don't think anyone is working on this. If you were to try this, you would setup a new CMake build variable (URHO3D_VULKAN), add new subdirectory to Source/Graphics subfolder, which is included in build only if the variable is set, and reimplement all the graphics classes (Graphics, Texture, VertexBuffer etc.) Be aware that the low-level graphics abstraction derives heavily from D3D9 and OpenGL2 states.

**Comment by TheComet93** [comment]
Just dropping this here as it might serve useful.
[](http://www.gamedev.net/page/resources/_/technical/opengl/vulkan-101-tutorial-r4408)

**Comment by 1vanK** [comment]
https://github.com/KhronosGroup/Vulkan-Hpp

**Comment by dragonCASTjosh** [comment]
@cadaver How much work would it be to move the abstraction to be based on D3D11 rather than D3D9 and would you allow such a commit. I think it would be beneficial as it would boost performance of D3D11 and GL3 alongside making it so something like vulkan likely have a performance benefit over API's.

**Comment by cadaver** [comment]
You would need to rework the Graphics class to work with D3D11-style state objects, make Materials contain them, and change the high-level rendering logic to work with the changed API. Maybe make shaders appear as if constant buffers are always in used (though on GLES2 they aren't). It's fairly substantial work. I'm not sure the gains are worth it, since in the end it's only about saving a few calls to the Graphics class; the D3D9 / GL style individual states are already being combined fairly effectively to D3D11 states, when using D3D11.

If you work on it, and make convincing before / after measurements, we would see if it's worth breaking the API.

In general I'm not sure if trying to significantly modernize Urho is worth it, unless you'd be prepared to rip out and rewrite significant parts of it.

**Comment by dragonCASTjosh** [comment]
Personally, I think modernising Urho is something that needs to happen at some point as its quickly being left behind.

If I find the time free I may look into it but I remember someone mentioned to me they was already experimenting with Vulkan in Urho but I don't remember who. As for focusing around a more modern style I noticed your personal engine turso3d uses a more modern renderer, what kind of performance difference do you see between that and Urho3D

**Comment by cadaver** [comment]
It's certainly already been left behind, and that's fine, as long as one understands what they're getting. Realistically speaking I'm not personally interested in trying to "keep up" with the big engines since the effort needed in relation to actually available developer power is huge. After all, this is a hobby for everyone involved. 

Something like internal representation of render states is only a band-aid or a very small piece in turning an engine truly modern, rather you'd need to upheave the API quite totally to utilize CPU cores better, and (like I've already mentioned before) most of the ease of use would go out of the window by that point. Properly programming for Vulkan or D3D12 is also a whole other, quite scary can of worms, and may require rethinking your rendering setup much beyond typical D3D11 use.

The sort of "dead end" that Urho finds itself in is IMO not particularly bad one, as we've got quite good feature parity across platforms, while still being much better performing than e.g. Ogre 1.x series, or even some versions of Unity. Ogre 2.x, on the other hand, has half of its platforms broken at this moment, because of the radical modernization approach (it can't even run on GLES2)

Turso3D is marginally more efficient, however it also has much missing features from the rendering (e.g. zones, light masks, light projectors), and less friendly, since the materials define raw constant buffer data blocks.

At this moment, my realistic view of what a manageable open source "engine" should be, is a low-level graphics API abstraction, with some lightweight tools on top of it, like octree and frustum culling, that would allow the user to write the actual renderer on top of it. Naturally that'd be much harder to use than Urho, since you wouldn't get any finalized renderer code out of the box. As an upside, you would be able to customize it for each project. It's likely I'm going to take Turso3D into that direction, when I have time to work on it.

**Comment by AlexDaniel** [comment]
Any updates/new thoughts on this?

**Comment by AlexDaniel** [comment]
Thanks bot, but it's still an issue.

**Comment by KondeU** [comment]
would you plan to support vulkan backend in 1.8.x version?

**Comment by tashaxing** [comment]
hope the Urlho3D official team could add vulkan support in later version release

**Comment by tigerpeng** [comment]
add openxr support too.
